### PR TITLE
#8671 Improving Table::findOrCreate()

### PR DESCRIPTION
**Pull request by burzum**
See https://github.com/cakephp/cakephp/issues/8671 for a description.

### CODE REVIEW

File: src/ORM/Table.php [hunk 1]
```diff
@@ -1235,6 +1235,25 @@ public function findOrCreate($search, callable $callback = null)
...
     /**
+     * Gets the query object for findOrCreate().
+     *
+     * @param array|\Cake\ORM\Query|string $search The criteria to find existing records by.
+     * @return \Cake\ORM\Query
+     */
+    protected function _getFindOrCreateQuery($search) {
+        if ($search instanceof Query) {
+            return $search;
+        } elseif (is_string($search)) {
```
Comment (dereuromark) [inline src/ORM/Table.php]: no else required

File: src/ORM/Table.php [hunk 2]
```diff
@@ -1235,6 +1235,25 @@ public function findOrCreate($search, callable $callback = null)
...
+            if (method_exists($this, $search)) {
+                return $this->{$search}();
+            } else {
```
Comment (dereuromark) [inline src/ORM/Table.php]: same

File: src/ORM/Table.php (hunk 1 above)
Comment (burzum) [inline src/ORM/Table.php]: Fixed. You also forgot the { which is in the same line. :) Fixed this as well.

File: src/ORM/Table.php (hunk 2 above)
Comment (burzum) [inline src/ORM/Table.php]: Fixed.

File: src/ORM/Table.php [hunk 3]
```diff
@@ -1235,6 +1235,26 @@ public function findOrCreate($search, callable $callback = null)
...
+     *
+     * @param array|\Cake\ORM\Query|string $search The criteria to find existing records by.
+     * @return \Cake\ORM\Query
+     */
+    protected function _getFindOrCreateQuery($search)
+    {
+        if ($search instanceof Query) {
+            return $search;
+        }
+        if (is_string($search)) {
```
Comment (markstory) [inline src/ORM/Table.php]: Could we get tests for this?

**Comment by markstory** [comment]
This should probably target `3.next` instead of `master` as it modifies an existing method.

File: tests/TestCase/ORM/TableTest.php [hunk 4]
```diff
@@ -5383,6 +5383,22 @@ function ($article) {
...
+        $query = $articles->find()->where(['author_id' => 2, 'title' => 'First Article']);
+        $article = $articles->findOrCreate($query);
+        $this->assertEquals('First Article', $article->title);
+        $this->assertEquals(2, $article->author_id);
+    }
+
+    /**
+     * Test that exceptions from the findOrCreate are thrown.
+     *
+     * @expectedException \InvalidArgumentException
```
Comment (dereuromark) [inline tests/TestCase/ORM/TableTest.php]: `@return void`

**Comment by dereuromark** [comment]
Awesome idea :+1:

**Comment by burzum** [comment]
Changing merge target. Damn Github for not allowing me to do this without a new PR...

https://github.com/cakephp/cakephp/pull/8707

**Pull request closed without merging**